Use mark element to indicate highlighted lines, ensure formatting when styles omitted, add Wrap Lines option

westonruter opened the pull request:
Fixes #79.
Fixes #78.

* Use `mark` element to indicate highlighted lines.
* Use CSS hack to force empty cell to show instead of line break.
* Prefix class names.
* Ensure Code block renders properly when no styles are loaded.
* Prevent flash of unstyled content by printing stylesheet before first Code block.
* Add Wrap Lines option, and ensure do not wrap by default for consistency.
* Improve code style.

With styles (note the lack of highlights and extra line breaks):

> ![image](https://user-images.githubusercontent.com/134745/80041867-68789700-84b2-11ea-8f04-8f9bb6dfee01.png)

Before without styles:

> ![image](https://user-images.githubusercontent.com/134745/80041935-98c03580-84b2-11ea-9828-a0a360050b67.png)

After without styles (with line spacing and highlights preserved):

> ![image](https://user-images.githubusercontent.com/134745/80042054-e046c180-84b2-11ea-995a-bb0d9728914c.png)

Note this is in the Twenty Twenty theme which has a default style of:

```css
mark {
	background-color: #eee;
	color: #222;
}
```

allejo commented on style.css:
```diff
@@ -1,29 +1,31 @@
+}
+
+.hljs.shcb-line-numbers .shcb-loc > span {
 	padding-left: 0.75em;
```
Add this https://github.com/westonruter/syntax-highlighting-code-block/commit/0de9d1058e901a7a227f42f3d1c3b37f0b9c46b9 and you'll knock out #78 also

westonruter replied:
Humm. I'm not seeing line wrapping at all anymore.

allejo replied:
I think we may want to address https://github.com/westonruter/syntax-highlighting-code-block/issues/78#issuecomment-617923236 at this point. If we add `white-space: pre-wrap` to each line, then we'll get the wrapping behavior and adding `display: table-cell` to the span will fix the indentation.

```patch
 .hljs.shcb-code-table > .shcb-loc {
 	display: table-row;
+	white-space: pre-wrap;
 	width: 100%;
 }
```

allejo replied:
At the moment, I'm having a hard time getting the `<code>` to respect the `<pre>`'s padding. This is because when I use `display: table`, it will disregard a parent's padding for `overflow` and only respect the parent's boundary. Resulting in this behavior:

![image](https://user-images.githubusercontent.com/1246453/80070485-df844e80-84f7-11ea-8b3d-a70c01f20866.png)

If you're open to changing our output from `<pre><code>` to `<pre><div><code>`, then that will let us have wrapping and overflow enabled very easily. Thoughts?

allejo replied:
Take a look at a503c31a79cbd2c8c4c34f47cb2e76c7de370808 and let me know what you think. Improvements welcome!

allejo commented on style.css:
```diff
@@ -1,29 +1,31 @@
 }
 
-.hljs.line-numbers .loc > span {
+.hljs.shcb-code-table > .shcb-loc > span::after {
+	content: "\200B"; /* Hack to force empty spans to render as table cells */
+}
```
This is the downside to use the CSS hack for rendering new lines, you lose the new line when you copy/paste. Since pseudo-elements are by definition not "selectable".

![2020-04-22 16-20-09 2020-04-22 16_21_03](https://user-images.githubusercontent.com/1246453/80042994-52200a80-84b5-11ea-9f70-c35368a463ca.gif)

westonruter replied:
Good find. What about 297d469?

allejo replied:
I like this much better! 👍

allejo commented:
If we do introduce the line wrapping option (which I just added to show how it'd work), I'm not sure how we want to handle that in Gutenberg. Do we want to modify the `PlainText`'s attributes to make it `nowrap`?

westonruter commented:
In regards to the wrapping I can see that regular block editor does line wrapping in the editor:

![image](https://user-images.githubusercontent.com/134745/80293791-7aa73f00-8717-11ea-925e-81f1984db4da.png)

But on the frontend no wrapping is done:

![image](https://user-images.githubusercontent.com/134745/80293794-8a268800-8717-11ea-921d-4b54b694d830.png)

Ah, but this is in Twenty Twenty. In Twenty Nineteen there _is_ line wrapping on the frontend:

![image](https://user-images.githubusercontent.com/134745/80293847-37010500-8718-11ea-8f24-fe54becee930.png)

So the inconsistent line wrapping seems like an inconsistency in the themes.

Therefore, if we add a line wrapping option we'd need not only have it _add_ line wrapping, but also to force line wrapping to be _removed_ if not enabled.

westonruter commented:
@allejo your changes fix the line wrapping in Twenty Nineteen:

![image](https://user-images.githubusercontent.com/134745/80293909-af67c600-8718-11ea-9bba-87eb6d4ead24.png)

And in Twenty Twenty (which does not do line wrapping on the frontend), it also looks good:

![image](https://user-images.githubusercontent.com/134745/80293921-c7d7e080-8718-11ea-95d3-fdbd48e578ad.png)

westonruter commented:
@allejo The Wrap Lines checkbox logic seems to be inverted:

![image](https://user-images.githubusercontent.com/134745/80294031-0cb04700-871a-11ea-81f3-9b11aa1edc57.png)

vs:

![image](https://user-images.githubusercontent.com/134745/80294032-189c0900-871a-11ea-876b-f89f2ed8e0e5.png)

westonruter commented:
> I'm not sure how we want to handle that in Gutenberg. Do we want to modify the `PlainText`'s attributes to make it `nowrap`?

Yes, I suppose so?

westonruter commented:
> If we do introduce the line wrapping option (which I just added to show how it'd work), I'm not sure how we want to handle that in Gutenberg. Do we want to modify the `PlainText`'s attributes to make it `nowrap`?

I just implemented this using CSS in b4c09ce. Maybe there's a better way.

Also fixed the inversion I noted in https://github.com/westonruter/syntax-highlighting-code-block/pull/80#issuecomment-619459563.

This is feeling solid.

allejo commented:
> I just implemented this using CSS in [b4c09ce](https://github.com/westonruter/syntax-highlighting-code-block/commit/b4c09ceb4d6862a155a37ce105ca7d2735a4d820). Maybe there's a better way.

I'm liking this implementation! Thanks for the fix

westonruter commented:
I'm seeing something strange that is blocking this from being merged. When attempting to make a change to a post with a Code block and hitting Update, I'm seeing this nonsensical error:

> ![image](https://user-images.githubusercontent.com/134745/80300261-f9b86980-874f-11ea-8c40-b9b52ebae644.png)

It's nonsensical because the JSON response is valid. The issue doesn't happen currently on `master`, but it doesn't on this branch. The issue happens when Gutenberg is deactivated as well.

westonruter commented:
I broke it in 28a8c20 somehow.

westonruter commented on style.css:
```diff
@@ -1,29 +1,44 @@
+
+.hljs.shcb-code-table {
 	display: table;
 	width: 100%;
```
@allejo Is the `width:100%` required? I'm finding in some themes, like Twenty Fourteen, it causes a bit of horizontal scrolling even when line wrapping is enabled. Can we remove it?

```suggestion
```

westonruter replied:
Humm. Seems like it is needed. If line numbers are turned off and the contents don't cause any wrapping, then the table is not wide enough:

![image](https://user-images.githubusercontent.com/134745/80312503-46c12d80-879a-11ea-85e6-159e34cd2030.png)

westonruter replied:
I found the problem with Twenty Fourteen: the `.hljs` element wasn't getting `box-sizing:border-box`, so the combination of `width:100%` and the `padding` resulted in an `overflow`. Fixed in e236851.
